src/dupin.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <Eigen/Dense>
+#include <algorithm>
+#include <iostream>
+#include <limits>
+#include <omp.h>
+#include <unordered_map>
+#include <vector>
+
+// DynamicProgramming class for dynamic programming based segmentation.
+class DynamicProgramming { // change name to Dynamic Programming
+private:
+  class upper_triangular_cost_matrix {
+  private:
+    std::vector<double> matrix;
+    int size;

```

Comment:
Change name to either dimension or length to help indicate it isn't the total size.